tests/subsys/fs/fs_api/src/test_multi_fs.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2020 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "test_fs.h"
+
+#define NUM_FS 2 // amount of file system
+#define TEST_FS_NAND1 "/NAND:"
+#define TEST_FS_NAND2 "/MMCBLOCK:"
+
+static struct test_fs_data test_data;
+
+static struct fs_mount_t test_fs_mnt_1 = {
+		.type = TEST_FS_1,
+		.mnt_point = TEST_FS_NAND1,
+		.fs_data = &test_data,
+};
+
+static struct fs_mount_t test_fs_mnt_2 = {
+		.type = TEST_FS_2,
+		.mnt_point = TEST_FS_NAND2,
+		.fs_data = &test_data,
+};
+
+static int test_fs_init(void)
+{
+	if (fs_register(TEST_FS_1, &temp_fs)) {
+		return -EINVAL;
+	}
+
+	if (fs_mount(&test_fs_mnt_1)) {
+		return -EINVAL;
+	}
+
+	if (fs_register(TEST_FS_2, &temp_fs)) {
+		return -EINVAL;
+	}
+
+	if (fs_mount(&test_fs_mnt_2)) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int test_fs_readmount(void)
+{
+	int ret;
+	int mnt_nbr = 0;
+	const char *mnt_name;
+
+	do {
+		ret = fs_readmount(&mnt_nbr, &mnt_name);
+		if (ret < 0) {
+			break;
+		}
+
+
+	} while (true);
+
+	if (mnt_nbr == NUM_FS) {
+		return 0;
+	}
+
+	return TC_FAIL;
+}
+
+static int test_fs_deinit(void)
+{
+	if (fs_unregister(TEST_FS_1, &temp_fs)) {
+		return -EINVAL;
+	}
+
+	if (fs_unmount(&test_fs_mnt_1)) {
+		return -EINVAL;
+	}
+
+	if (fs_unregister(TEST_FS_2, &temp_fs)) {
+		return -EINVAL;
+	}
+
+	if (fs_unmount(&test_fs_mnt_2)) {
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int test_fs_unsupported(void)
+{
+	if (fs_register(UNSUPPORTED_FS, &temp_fs) == 0) {
```

Comment:
I think you should be using `FS_TYPE_END` here explicitly, and in the `.type` initialization of `test_fs_mount_unsupported_fs`.  It's that value, representing the maximum number of file systems, that isn't allowed.

This reveals a weakness in the common code.  `enum fs_type` provides "unique" identifiers for file systems.  It should allow for externally-defined file systems and it should not be used to determine the size of `fs_map.c` in `fs.c`.  Rather that should be defined by a Kconfig variable based on the number of file system types required by an application.